Add the StorageLite feature

davidsaada opened the pull request:
### Description
StorageLite is a storage solution, providing a key value store like API of set/get data or a reduced POSIX API of open-write-close and open-read-close.

Design & documentation by @offirko
General and file system tests by @theamirocohen  

[Handbook documentation PR](https://github.com/ARMmbed/Handbook/pull/512)

This PR depends on the following open PRs:
[#6757](https://github.com/ARMmbed/mbed-os/pull/6757) - Buffered block device
[#6559](https://github.com/ARMmbed/mbed-os/pull/6559) - Flash simulator block device
[#6480](https://github.com/ARMmbed/mbed-os/pull/6480) - NVStore allocate_key API
[#6642](https://github.com/ARMmbed/mbed-os/pull/6642) - Device key


### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] New target
    [x] Feature
    [ ] Breaking change

Commit:
Add StorageLite feature

davidsaada commented:
@dannybenor @geky

0xc0170 commented:
Let's make sure those dependencies are moving forward

geky commented on features/device_key/source/DeviceKey.h~HEAD:
```diff
@@ -0,0 +1,142 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_DEVICEKEY_H
```
I think this file got added by accident? (the ~HEAD part)

dannybenor replied:
@geky Your view is clear and you presented it several times.
Review of this PR should be done in terms of code quality and not feature definitions. There are good reasons to keep two different APIs for storage but this is not the place to discuss this.
Will appreciate your help on code quality review

geky replied:
I hope you will find the review itself is constructive.

However this is the correct place to note why a PR should not be merged, otherwise the PR will be merged.

davidsaada replied:
The Head file is part of the dependent DeviceKey PR and should be handled there. @yossi2le are you aware of this?

yossi2le replied:
@davidsaada  The DeviceKey PR looks fine, I have no idea why this file is showing on this PR.

davidsaada replied:
Maybe it was accidentally added while rebasing. Will check.

geky replied:
No rush, you'll need to rebase when dependencies go in anyways

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+    : FileSystem(name), _stlite(stlite), _flags(flags), _bd(0)
+{
+    MBED_ASSERT(stlite);
+    mount(_bd);
```
I had to remove this mount call because the _bd pointer was NULL and the init assert would fail.

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+            int ret = _stlite->file_exists((const uint8_t *) path, strlen(path) + 1);
+            // Fail if file exists
+            if (ret == STORAGELITE_SUCCESS) {
+                return EEXIST;
```
Should this be negative?

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+    }
+
+    // We don't support opening for read-write, nor for appending
+    if ((flags & O_RDWR) || (flags & O_APPEND)) {
```
Yay! Glad to see full flag support :+1:

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+int StorageLiteFS::file_open(mbed::fs_file_t *file, const char *path, int flags)
+{
+    file_handle_data_t *handle = new file_handle_data_t;
+    if (!handle) {
```
This won't come back as NULL, it will assert. I believe you need [nothrow](http://www.cplusplus.com/reference/new/nothrow/):
``` cpp
file_handle_data_t *handle = new (std::nothrow) file_handle_data_t;
```

davidsaada replied:
Fixed.

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+        }
+    }
+
+    handle->file_name = new char[strlen(path) + 1];
```
needs nothrow if you want to return an error.

offirko replied:
would consider verifying strlen(path) is smaller than max file name size to prevent memory corruption or security problem

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+        return -ENOSYS;
+    }
+
+    int ret = _stlite->set((const uint8_t *) handle->file_name, strlen(handle->file_name) + 1,
```
Why `strlen(handle->file_name) + 1`, does the null terminator need to be stored?

davidsaada replied:
No special reason. Can remove it.

davidsaada replied:
Null terminator removed.

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+
+int StorageLiteFS::file_sync(mbed::fs_file_t file)
+{
+    return -ENOSYS;
```
This shouldn't error. If the file is always in sync with the underlying storage this should be a no-op that returns success.

davidsaada replied:
Correct. Will fix it.

davidsaada replied:
Fixed. Actually removed the implementation altogether, as the base class function just returns 0.

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+
+off_t StorageLiteFS::file_seek(mbed::fs_file_t file, off_t offset, int whence)
+{
+    return 0;
```
Hmm, shouldn't this error with ENOSYS?

davidsaada replied:
Would agree, but please notice that the function doesn't return an error code, but an offset.

geky replied:
off_t is signed for this reason and returning ENOSYS from seek is expected. (I know it's not obvious from the type's name, that comes from POSIX).

davidsaada replied:
Fixed.

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+
+    // We only support one (correct) read/write operation after open
+    if (((handle->open_flags & 1) == O_WRONLY) || !handle->file_name) {
+        return -ENOSYS;
```
This is either 0 or the file size after a write isn't it? Do we not keep track of the file size?

davidsaada replied:
It should either be the file size (if the operation succeeds) or a negative value on failure. We don't need to keep track of the file size - it is only relevant at the time of the successful operation.

geky replied:
Ah ok, it's probably fine then

geky commented on features/filesystem/storagelite/source/StorageLiteFS.h:
```diff
@@ -0,0 +1,268 @@
+     * @return                              none.
+     */
+    StorageLiteFS(const char *name = 0, StorageLite *stlite = 0,
+                  uint32_t flags = StorageLite::encrypt_flag);
```
Why not have these flags per file?

davidsaada replied:
StorageLite flags are proprietary. Having these implemented on a per file basis (in the POSIX open API) would make us add them to the standard flags. This would be misleading as other file systems don't support these flags.

geky replied:
Not really a problem if it doesn't conflict with standard Mbed OS flags. The proprietary flags can be provided with a prefix to set them apart.

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+        case STORAGELITE_MAX_FILES_REACHED:
+            return -ENFILE;
+        case STORAGELITE_NOT_SUPPORTED:
+            return -ENOSYS;
```
Actually, why do we have storagelite specific errors and not use errno codes in the storagelite layer? It doesn't look like there's much lost information. This would be a small code size improvement getting rid of this switch/case layer.

davidsaada replied:
This would initially make sense, but it is actually very problematic: 
First of all, as you can see from the code, few of them don't make too much sense. The uninitialized error, for instance, has no equivalent in the errno set. So is the "OS error" (implying an internal error). But that's just the least of the problems. It looks like the entire handing of errno based codes in mbed-os is quite flawed. There are conflicts between the numerical values of these codes between the different compilers, and between those and the ones exported by mbed-os itself (in `mbed_retarget.h` include file). Would raise an issue about it, but as @SenRamakri is about to introduce a mechanism to replace those, it seems to become irrelevant soon.

davidsaada replied:
Just to add - our FS test initially included an errno check that failed when compiled by the ARM compiler, as the errno numbers didn't match the ones in mbed-os. I believe any test that would try checking errno values would fail the same way.

geky commented on features/filesystem/storagelite/mbed_lib.json:
```diff
@@ -0,0 +1,16 @@
+    "macros": ["MBEDTLS_CIPHER_MODE_CTR", "MBEDTLS_CMAC_C", "MBEDTLS_ENTROPY_C"],
+    "config": {
+        "enabled": {
+            "macro_name": "STORAGELITE_ENABLED",
```
Why do we have a STORAGELITE_ENABLED flag, shouldn't it always be enabled?

davidsaada replied:
Uh, maybe some users don't need this feature? It still consumes some code and RAM space.

geky replied:
It should be garbage collected by the compiler then. We usually only have these flags for libraries that can't be garbage collected by the compiler (ie rtos and lwip due to statically allocated RAM regions). StorageLite is just logic, so it should be easily removable by the compiler.

davidsaada replied:
Makes sense. Will remove from the configuration file, but not from the code, as I see that some of the build failures are on boards that don't have NVStore, and I'll change this define to depend on NVStore enabled one.

geky replied:
Ah good idea :+1:

geky commented on features/filesystem/storagelite/mbed_lib.json:
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "storagelite",
+    "macros": ["MBEDTLS_CIPHER_MODE_CTR", "MBEDTLS_CMAC_C", "MBEDTLS_ENTROPY_C"],
```
Just double checking, is this the correct way to enable mbedtls features? @cmonr who's a TLS contact?

davidsaada replied:
I think it's the correct way. However, the `MBEDTLS_ENTROPY_C` flag was wrongly used and removed now (checking it exists now instead of wrongfully forcing it).

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     *
+     * @returns Number of files.
+     */
+    size_t get_max_files() const;
```
Hmm, if this isn't a filesystem API, should these really be called files?

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     *          STORAGELITE_BAD_VALUE         Bad value in any of the parameters.
+     *          STORAGELITE_BUFF_TOO_SMALL    Not enough memory in user buffer.
+     */
+    int get(const uint8_t *file_name, size_t file_name_size, void *data_buf,
```
Could we add an overload for a null-terminated string? Most users probably just want to use a C string:

``` cpp
int StorageLite::get(const char *file_name, void *data_buf, size_t data_buf_size, size_t& actual_data_size)
{
    return get(file_name, strlen(file_name), data_buf_size, actual_data_size);
}
```

davidsaada replied:
I agree. Initial requirement was to have an opaque "blob name", hence this type.

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     *          STORAGELITE_BUFF_TOO_SMALL    Not enough memory in user buffer.
+     */
+    int get(const uint8_t *file_name, size_t file_name_size, void *data_buf,
+            size_t data_buf_size, size_t& actual_data_size);
```
actual_data_size should be a pointer in this situation to make the intention clear at the call site

davidsaada replied:
Got to disagree here. This code is implemented in C++, so references should be used. Pointers fit to the C like programming scheme.

geky replied:
We had a discussion about this a while back and agreed to use C pointers. I'm not sure where it is documented, @0xc0170?

For example, consider [line 709](https://github.com/ARMmbed/mbed-os/pull/6915/files/0a55174d36988ad2cfa9bbdab567400d8cf8711c#diff-340d6de032717e89f42100e4c0e8da0aR709):
``` cpp
os_ret = nvstore.set(header.nvstore_key, cmac_size, header.cmac);
```

Does this call modify header.cmac? What about cmac_size?

I understand your point about being consistent with other C++ code, however that has not been mbed's goal from what I understand.

davidsaada replied:
Well, I believe that the user should check the meaning of each parameter (in documentation for instance), where the parameter being modified or not is just one aspect. I personally find references very useful and dislike the usage of asterisks and ampersands throughout the function code and calls. My stand here is that you if you use C++, you can use any of its features apart from the ones strictly forbidden by code rules (exceptions, STL etc.). Otherwise you'd be just be programming in C.

davidsaada replied:
And BTW, your example partially proves my point: header.cmac is an array, and as such it is a pointer. Even if you were using pointers for modified parameters, you wouldn't be noticing that from the function call. For that you should be looking at the function prototype and see that the parameter is of `const void *` type.

geky replied:
It's generally agreed upon that arrays in C/C++ aren't well integrated into the language. Though without references, you can still be confident that the address where `header.cmac` points won't change.

I do want to pull in @0xc0170 on this, at the end of the day consistency is more important than either of our opinions and @0xc0170 generally makes the final call on style for this reason.

davidsaada replied:
If consistency is what you seek, please observe too examples where references are used and not pointers (and I guess there are more):
[LoRaWan get_plan function](https://github.com/ARMmbed/mbed-os/blob/e1a63f9046515df8d7f66ce251c3749812506bf6/features/lorawan/lorastack/mac/LoRaMacChannelPlan.h#L84)
(Note that in this case, the `set_plan` function uses the `const` directive, meaning that there's no way telling if it's a returning value or not from the function call, you just need to look at the header).
[NVstore get function](https://github.com/ARMmbed/mbed-os/blob/e1a63f9046515df8d7f66ce251c3749812506bf6/features/nvstore/source/nvstore.h#L127)
(just one example out of many in the NVStore code).

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     *          STORAGELITE_DATA_CORRUPT      Data is corrupt.
+     *          STORAGELITE_BAD_VALUE         Bad value in any of the parameters.
+     */
+    int get_file_size(const uint8_t *file_name, size_t file_name_size, size_t& actual_data_size);
```
returning args should be pointers

davidsaada replied:
Plz see my C vs. C++ comment.

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     *          STORAGELITE_DATA_CORRUPT      Data is corrupt.
+     *          STORAGELITE_BAD_VALUE         Bad value in any of the parameters.
+     */
+    int get_file_flags(const uint8_t *file_name, size_t file_name_size, uint32_t& flags);
```
returning args should be pointers

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     *          STORAGELITE_READ_ERROR    Physical error reading data.
+     *          STORAGELITE_WRITE_ERROR   Physical error writing data (on recovery).
+     */
+    int init(BlockDevice *bd, size_t max_files = STORAGELITE_MAX_FILES);
```
If this is a new API I would actually discourage providing the block device in init, I believe that was a mistake. Instead I would take it in the constructor and let bd default to NULL here.

+1 for the handling of max_files here

davidsaada replied:
This followed the FileSystem base class. I agree it is awkward and it made me add some code pirouettes, specially at init and reset. However, changing this would be a breaking change for all current file systems (FAT, LittleFS), and would require changes in legacy code (PAL code in client is one example). So, regretfully had to follow current base class.

geky replied:
Code pirouettes is an amazing word choice. Also I understand the problem now. :+1: 

I agree this should stay as is. We should see if we can change this in the FileSystem API in the future.

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     *
+     * @returns StorageLite free size.
+     */
+    size_t free_size() const;
```
Do you need free_size if you have size and can get the block device's size?

If you have these it shouldn't be hard to implement statvfs in StorageLiteFS.

davidsaada replied:
`free_size` is based on the current pointer to the next written record in the BD. One can't calculate it from the BD size (which is static). I thought 
Regarding `statvfs` - yes, was wondering whether to implement it or not. Most of the info in this structure has no meaning for StorageLite (like "Number of free blocks for unprivileged users"). Maybe will implement it anyway. No big deal.

geky replied:
Oh interesting, is there a way to get the theoretical size if a gc were to happen?

I was thinking this was a representation of how "full" the filesystem was. Usually this is what users expect the "size" to be. Maybe these APIs are too confusing? Are they useful for something?

The biggest use of statvfs is "how full" the filesystem is, mostly for logging purposes.

davidsaada replied:
It is used for that purpose ("how full"). Of course it's misleading, because it doesn't take future metadata and padding into account, just the "raw" free size.

davidsaada replied:
Anyway, `statvfs` has been implemented.

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     * @returns STORAGELITE_SUCCESS           Success.
+     *          STORAGELITE_BAD_VALUE         Bad area parameter.
+     */
+    int get_area_params(uint8_t area, uint32_t& address, size_t& size);
```
Out of curiousity, what's the value of this API?

davidsaada replied:
Ugh. Will be removed. This was copied from NVStore, and while it had its value there, it has no value here.

davidsaada replied:
Removed.

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+    static const uint32_t update_backup_flag;
+
+    static const size_t max_name_size;
+    static const size_t max_data_size;
```
As constants, these should be all caps. Also should should they be in an enum?

davidsaada replied:
Are you referring to the flags? If so, then s we can use a combination of these, they should certainly not be defined as enums. As for them being all caps, I guess it would be true if they were defined with `#define`, but not sure whether it's the case if they are defined as `const` (maybe it's another C vs C++ argument). Do we have any rules for that? If so, will change these according to these rules.

geky replied:
Fair point for enums.

In regards to name style: 
@0xc0170? My vote is all caps for any thing that is effectively a constant. This also makes it easier to switch between #defines and const variables if needed.

geky commented on features/filesystem/storagelite/source/StorageLite.h:
```diff
@@ -0,0 +1,559 @@
+     *          STORAGELITE_BUFF_TOO_SMALL    Not enough memory in user buffer.
+     */
+    int get_first_file(uint8_t *file_name, size_t file_name_buf_size,
+                       size_t& actual_file_name_size, uint32_t& handle);
```
I think this would be better abstracted as a C++ class representing the iteration (like the Dir class, though it could be a different class, maybe StorageLiteIterator?). In the future it could fit into some nice C++11 features such as the for each loop.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+// Internal flags - flags added by us (like deleted flag)
+static const uint32_t all_rec_flags_mask                    = 0xFFF00000UL;
+static const uint32_t ext_rec_flags_mask                    = 0xFF000000UL;
+static const uint32_t rec_data_size_mask                    = 0x000FFFFFUL;
```
Constants should be all-caps

davidsaada replied:
See my comment above.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+static const uint32_t min_prog_align_size = 16;
+
+static const unsigned char *enc_key_salt  = (const unsigned char *) "Encrypt key salt";
+static const unsigned char *auth_key_salt = (const unsigned char *) "Auth. key salt  ";
```
Hmm, should this be configurable so users can choose their own salt?

davidsaada replied:
Don't think so. Why lay this burden n the user if used for internal purposes? But I'll change the salt to include "StorageLite", so it won't conflict with other user salt strings.

geky replied:
If its a config option, users wouldn't _need_ to know about it. But it would still be available for an application that wants the additional security.

Unsure though, I'll leave it up to you.

davidsaada replied:
Kept as is, just changed the salt names to include StorageLite.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+static const uint32_t enc_block_size = 16;
+
+static const uint16_t master_rec_format_rev = 0;
+static const uint8_t *master_rec_file_name = (const uint8_t *) "master_rec";
```
Thoughts on making this "storagelite" or something so it shows up if a user hexdumps a piece of storage?

davidsaada replied:
Good idea.

davidsaada replied:
Changed.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+} master_record_data_t;
+
+const size_t StorageLite::max_name_size = 1024;
+const size_t StorageLite::max_data_size = 256 * 1024;
```
These should be provided as public constants similar to NAME_MAX in POSIX, that way users can know how much memory they need to allocate to load names.

davidsaada replied:
Right. Will change.

davidsaada replied:
Fixed.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+}
+
+int StorageLite::get_file_flags(const uint8_t *file_name, size_t file_name_size,
+                                uint32_t& flags)
```
Between `get_file_size`, `get_file_flags`, and `file_exists`, could these not all be wrapped up into one function that gets all the metadata of a file? This could be wrapped up in a struct:
``` cpp
struct storagelite_info info;
stlite->stat(&info);
```

This would all the user to get all of the metadata with one access to storage.

Also we could add the metadata as an overload to the basic get function, allowing it to all be captured at once.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+    // This handles the case that init function is called by more than one thread concurrently.
+    // Only the one who gets the value of 1 in _init_attempts_val will proceed, while others will
+    // wait until init is finished.
+    init_attempts_val = core_util_atomic_incr_u32(&_init_attempts, 1);
```
Doesn't grabbing the storage lite's mutex avoid this race condition?

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+
+    NVStore& nvstore = NVStore::get_instance();
+    os_ret = nvstore.init();
+    MBED_ASSERT(!os_ret);
```
Should this error actually be propogated upwards?

davidsaada replied:
Believe that all errors at init stage should be considered fatal.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+    }
+
+    if (!_mutex) {
+        _mutex = new PlatformMutex;
```
Why dynamically allocate this mutex?

davidsaada replied:
When it comes to "non trivial" member variables, I believe they should be dynamically allocated as part of the init phase (delaying the memory consumption as much as possible).

geky replied:
Interesting, I view the mutex as a "trivial" member variable. The mutex doesn't have a buffer or anything, on some platforms it could be as small as an int.

davidsaada replied:
There was initially another reason, which eventually didn't apply here (was relevant in NVStore): Few of the tests used thread killing. Now mutexes (when taken) don't like this and sometimes get stuck when it happens. Deleting the mutex was a good workaround for this problem.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+        _mutex = new PlatformMutex;
+    }
+
+    bd->init();
```
Looks like we missed an error

davidsaada replied:
Right. will fix.

davidsaada replied:
Fixed.

geky commented on features/filesystem/storagelite/source/StorageLite.cpp:
```diff
@@ -0,0 +1,1846 @@
+    // read_record and write_record code much more complex (as we'll have to read/write chunks
+    // of the data along with the header). If this is the case, use the buffered block device
+    // to reduce the underlying BD read and program sizes
+    if (bd->get_program_size() <= sizeof(record_header_t)) {
```
In addition, `sizeof(record_header_t)` should be a multiple of the `program_size`, otherwise alignment becomes a problem.

I think it may just be simpler to check if `bd->get_program_size() == 1`, unless `record_header_t` is a very common program size.

davidsaada replied:
Agree about checking the header size as a multiple of the program size. Off topic, I think block devices should make sure read/program/erase sizes are powers of 2. This would make that check redundant.

geky replied:
I'm not sure I understand, that only helps here if `sizeof(record_header_t)` is also a power of 2 doesn't it?

davidsaada replied:
I should also point that out in the comments - header is 32 bytes, and any change should keep it a power of 2.

geky replied:
IMO this is a pretty small optimization when there's already block devices without non-power of 2 blocks (dataflash for ECC). But we can add a util block device for conversion to power-of-2 if necessary, at a storage size tradeoff.

davidsaada replied:
Anyway fixed to check whether it's a multiple.

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+
+Uses: keeping credentials and cryptographic keys, application settings, application states and small application data to be sent to servers.
+
+Hardware: designed for external NOR flash, it can work also on internal flash and with additional block devices on SD cards.
```
We should probably mention the user will need the FlashSimBlockDevice to use SD cards. Did we measure the performance on SD cards?

deepikabhavnani replied:
FlashSimBlockDevice  is simulation of flash over RAM, SD is already flash memory.
 
> it can work also on internal flash and with additional block devices on SD cards.

Please clarify this. Which additional block device and why

davidsaada replied:
Good point @geky. Will add it to documentation. Got a few performance numbers, can send them later today. Maybe we should run your benchmark on this as well.
@deepikabahavani The Flash simulator adaptor applies to any block device not having flash attributes (i.e. erasing). SD is such a device thus requiring the flash simulator adaptor.

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+  - Integrity through file authentication.
+  - Confidentiality through file encryption.
+  - Rollback protection.
+- Wear level: StorageLite stores files in a sequential order. This concept uses the entire external memory space. It is optimized for NOR flash that supports a limited number of erase per sector.
```
We can also suggest using a SlicingBlockDevice or MBRBlockDevice to not use the entire external memory space since one feature StorageLite is bring is small storage usage.

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+  - Rollback protection.
+- Wear level: StorageLite stores files in a sequential order. This concept uses the entire external memory space. It is optimized for NOR flash that supports a limited number of erase per sector.
+- Backup: built-in backup and factory reset support.
+- Low RAM memory footprint: indexing each file uses 8 bytes of RAM.
```
Maybe better wording is "indexing using 8 bytes of RAM per file." We want to hint that users can optimize RAM consumption by reducing the number of files.

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+- Backup: built-in backup and factory reset support.
+- Low RAM memory footprint: indexing each file uses 8 bytes of RAM.
+
+Please see the [StorageLite design document](./StorageLiteDesign.md) for a detailed design description of StorageLite.
```
Is this going to come in with this pr?

davidsaada replied:
Will update the link once the handbook PR merges.

geky replied:
Ah! Didn't see the handbook PR, will look into that 👍

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+
+StorageLite is optimized for NOR flash memory. It is based on dividing the allocated memory storage to areas: active and standby. Data is written to the active area until it becomes full. When the active area becomes full, an internal garbage collection mechanism moves only the updated relevant files to the standby area and switches between the active and standby areas. The new standby area is th ...
+
+This concept is ideal for low wear leveling of the memory, but it comes with a price of using half of the external memory size.
```
I don't think we should say half here. Maybe a quarter? Otherwise users may measure out the minimum amount of storage needed and slice a block device to fit 2*(files+metadata). Unfortunately, in this case StorageLite will be forced to perform a garbage collect every write.

davidsaada replied:
maybe "about half" would be better here.

geky replied:
Maybe "a portion" or "a subset"?

Oh, or "but it comes with a price of not being able to use more than half of the external memory size".

I just don't want to encourage users to use exactly half, thinking that's the optimal amount of storage.

geky replied:
Some quick maths:

Writing `n` byte files to StorageLite with `d` dynamic files and `s` static files gives us this equation:

![n](https://latex.codecogs.com/svg.latex?n%20&plus;%20n%5Cfrac%7B%20s%7D%7Bd&plus;1%7D)

If we let `r` be the ratio of static space to the size of storage, we can build these guys:

![s](https://latex.codecogs.com/svg.latex?s%20%3D%202r%5Cfrac%7Bsize%7D%7Bn%7D)
![d](https://latex.codecogs.com/svg.latex?d%20%3D%20%281-2r%29%5Cfrac%7Bsize%7D%7Bn%7D)

This gives us a nice formula for the cost of writing `n` byte files with a ratio `r` of how full the storage is:

![n](https://latex.codecogs.com/svg.latex?n%20&plus;%20n%5Cfrac%7B2r%5Cfrac%7Bsize%7D%7Bn%7D%7D%7B%281-2r%29%5Cfrac%7Bsize%7D%7Bn%7D&plus;1%7D)

Lets assume 100 bytes files on a 1MB block device, we can plot this:
![image](https://user-images.githubusercontent.com/1075160/40135144-2430b580-590a-11e8-918b-2b889b914491.png)

So at 25% usage, we're seeing on average 200 B written per file, and at 37.5% usage we're seeing on average 400 B written per file.

I think this is fine, compared to every other resource flash is massive. But we should probably encourage users to stick to around 37.5% flash usage. (Feel free to steel these maths for documentation).

offirko replied:
The analysis is good, but I don't believe we should get into this detailed level of analysis or recommendation. The main point to emphasize to the user was that StorageLite reduces the storage potential by half to begin with. "but it comes with a price of not being able to use more than half of the external memory size" sounds great to me (-:

geky replied:
May be useful to have in the design documentation, but up to you :+1:

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+
+#### APIs
+
+You can either use StorageLite with set- and get-based APIs or it can be defined as StorageLiteFS and work with a limited POSIX open-write or read-close based API. (Only a **single** write or read operation is allowed.)
```
``` diff
- Only a **single** write or read operation is allowed.
+ Only a **single** write or read operation is allowed per file.
```

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+- `file_exist`: verifies that file exists and authenticates it.
+- `get_file_flags`: returns files supported features (such as rollback protection, encryption and factory reset).
+- `get_first/next`: provides the ability to iterate through existing files, starting from the first one.
+- `factory_reset`: returns storage to contain only files that are backed up and returns those files to their backup version.
```
Should we mention `reset` here?

Out of curiousity does `factory_reset` fall back to `reset` if there's no factory? Should these functions be combined?

davidsaada replied:
Reset equals format and it should definitely be mentioned. And factory reset indeed falls back to reset if no factory backup file has been created. These shouldn't be combined as they are two totally different features fulfilling different requirements.

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+- Write: Writing data to the file associated with the opened file descriptor (replacing existing data - if exists). You can only call write once.
+- Read: reads data from the file associated with the open file descriptor. Read can be called only once.
+- Remove: deletes a file name from the file system.
+- Reformat: clears a file system, resulting in an empty and mounted file system.
```
We can probably just link to the FileSystem reference (or copy and paste it here):
https://os.mbed.com/docs/v5.8/mbed-os-api-doxy/classmbed_1_1_file_system.html#a387bceba0e4696753a16bab3a501f8a4

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+- `init`: binds StorageLite to a specific block device and constructs and initializes StorageLite module elements.
+- `deinit`: deinitialize the StorageLite module.
+- `set`: saves data into persistent storage by file name.
+- `get`: retrieves file data from storage after authenticating it.
```
If we're introducing a new storage API, we should provide an abstract interface for it (like the FileSystem class). Maybe KVStore? Feel free to come up with your own name.

The new API should be provided in its own abstract class (no members) and implemented by StorageLite. We'll need an mbed-os-example-kvstore (like mbed-os-example-filesystem), and in the future KVStore tests (we don't have portable filesystem tests so this may all have to come later).

Additionally we should provide KVStore implementations for all FileSystems that can support the new API (maybe this should be in the FileSystem abstract class itself?).

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+
+Hardware: designed for external NOR flash, it can work also on internal flash and with additional block devices on SD cards.
+
+StorageLite is a lightweight file system storage module for files in external memory. Files are stored and accessed by file-name. StorageLite has the following attributes:
```
If StorageLite is not a FileSystem should we still call its entries files?

davidsaada replied:
Can't think of a better name. Entries can be OK, but it can also be misleading.

geky replied:
Maybe "values" or "data" or "blobs" or "elements". Up to you.

geky commented on features/filesystem/storagelite/README.md:
```diff
@@ -0,0 +1,92 @@
+- Backup: built-in backup and factory reset support.
+- Low RAM memory footprint: indexing each file uses 8 bytes of RAM.
+
+Please see the [StorageLite design document](./StorageLiteDesign.md) for a detailed design description of StorageLite.
```
@offirko, @davidsaada, I realize this is coming late, so not blocking this pr, but could we also have a SPEC.md to cover just the layout of data on disk? Similar to the SPEC.md in LittleFS:
https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/littlefs/SPEC.md

This would help with forward compatibility and tooling.

offirko replied:
Hi, layout of data on disk is described in the design part of the handbook documentation.
https://github.com/ARMmbed/Handbook/pull/512/commits/85f1c15ee11edd507305129ee85df308c056da74#diff-c3ac614f9d51ef7de09c0f38a05b4e75

geky commented on features/filesystem/storagelite/source/StorageLiteFS.h:
```diff
@@ -0,0 +1,268 @@
+/**
+ * StorageLiteFS, FileSystem interface for the StorageLite feature
+ */
+class StorageLiteFS : public mbed::FileSystem {
```
With an abstract KVStore interface, this adapter does not have to be StorageLite specific and provide the FileSystem API for any KV implementation.

geky commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+
+int StorageLiteFS::dir_open(mbed::fs_dir_t *dir, const char *path)
+{
+    return -ENOSYS;
```
I think this is already the default implementation for dir operations in the FileSystem API. If you omit the functions from the class they should return ENOSYS by default.

geky commented on features/filesystem/storagelite/TESTS/storagelite/fs_tests/main.cpp:
```diff
@@ -0,0 +1,669 @@
+#ifdef TEST_SD
+#include "SDBlockDevice.h"
+#endif
+#endif
```
I don't think you need these includes. You should be able to use the include logic from the filesystem tests:
https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/TESTS/filesystem/files/main.cpp#L59-L68

davidsaada replied:
They are there for our local testing (when local test is defined, not defined by default).

geky commented on features/filesystem/storagelite/TESTS/storagelite/fs_tests/main.cpp:
```diff
@@ -0,0 +1,669 @@
+
+
+#ifdef TEST_SPIF
+#ifdef TARGET_K82F
```
This should be using the declaration logic from the filesystem tests:
https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/TESTS/filesystem/files/main.cpp#L59-L68

That way it can run in our CI.

geky commented on features/filesystem/storagelite/TESTS/storagelite/fs_tests/main.cpp:
```diff
@@ -0,0 +1,669 @@
+    Case("StorageLiteFS_fread_with_fopen_w_mode", StorageLiteFS_fread_with_fopen_w_mode),
+    Case("StorageLiteFS_fread_closed_file", StorageLiteFS_fread_closed_file),
+    Case("StorageLiteFS_fread_twice_from_file", StorageLiteFS_fread_twice_from_file),
+    Case("StorageLiteFS_fread_twice_separated_with_fclose", StorageLiteFS_fread_twice_separated_with_fclose),
```
Good amount of tests! We need these on all filesystems!

geky commented on features/filesystem/storagelite/TESTS/storagelite/general_tests/main.cpp:
```diff
@@ -0,0 +1,1493 @@
+static const size_t bd_read_size           = 1;
+
+StorageLite *stlite = NULL;
+HeapBlockDevice bd(bd_size, bd_read_size, bd_prog_size, bd_erase_size);
```
We also have an MBED_TEST_SIM_BLOCKDEVICE which I think is useful here. Otherwise the CI will try to run this on platforms with very little RAM:
https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/TESTS/filesystem_recovery/wear_leveling/main.cpp#L48-L58

davidsaada replied:
Not sure whether you saw the beginning of this file, failing the support of this test if `MBED_TEST_BLOCKDEVICE` is not present. This should prevent the execution of this test on low memory boards.

davidsaada replied:
Just to add: The test currently doesn't use the BD exported by `MBED_TEST_BLOCKDEVICE`. Reason is that I wanted to have this white box test on different types of BDs (in term of sizes). I'll try adding this BD to the list of tested BDs.

geky commented on features/filesystem/storagelite/TESTS/storagelite/whitebox/main.cpp:
```diff
@@ -0,0 +1,670 @@
+static const char *file3_val1 = "Data value of file 3 is the following";
+static const char *file4_name = "This is the name of file4";
+static const char *file4_val1 = "Is this the value of file 4?";
+static const char *file4_val2 = "What the hell is the value of file 4, god damn it!";
```
:laughing:

kjbracey replied:
General tip, even though this is a test - this is blowing static RAM probably unnecessarily - 4 words of RAM for 4 non-constant pointers.

Suggest `static const char * const file4_name` or `static const char file4_name[]`.

The presence of the pointer `const` qualifier tends to fool one into thinking that the object is `const`.

geky replied:
It's in a test and probably optimized out anyways,

kjbracey replied:
Indeed, but it's a common pattern I see, so I like to raise awareness.

And you're right, as these are `static` the compiler could well optimise the pointer storage away - if it's analysing enough of the source file at a time to be certain they're not modified.

davidsaada replied:
Good catch @kjbracey-arm. Adding that extra `const` actually reduced code size, at least in the ARM GCC compiler.

geky commented on features/filesystem/storagelite/TESTS/storagelite/whitebox/main.cpp:
```diff
@@ -0,0 +1,670 @@
+/*
```
Out of curiousity what does "whitebox" mean?

davidsaada replied:
White box is a testing term meaning that the test is aware of the internal coding logic (opposite from black box). Here the test code is aware of edge cases (BD and erase sizes for example) and different code paths.

geky replied:
Ah! That makes perfect sense, thanks

geky commented on features/filesystem/storagelite/TESTS/storagelite/whitebox/main.cpp:
```diff
@@ -0,0 +1,670 @@
+    Case("StorageLite: White box test",     white_box_test,    greentea_failure_handler),
+    Case("StorageLite: White box FS test",  white_box_fs_test, greentea_failure_handler),
+    Case("StorageLite: Multiple set test",  multi_set_test,    greentea_failure_handler),
+    Case("StorageLite: Error inject test",  error_inject_test, greentea_failure_handler),
```
This is a great number of tests, but I think we're missing a couple important ones:
- Simulated power-resilience testing: [example](https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/TESTS/filesystem_recovery/resilience/main.cpp)
- Greentea-driven power-resilience testing: [example](https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/TESTS/filesystem_recovery/resilience_functional/main.cpp)
- Simulated wear-level testing: [example](https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/TESTS/filesystem_recovery/wear_leveling/main.cpp)

Note: these depend on some common files:
https://github.com/ARMmbed/mbed-os/tree/master/features/filesystem/littlefs/TESTS/COMMON

These already exist on top of the filesystem API, so it shouldn't be too difficult to adopt them here. Until then it would  be a risk to advertise power resilience.

davidsaada replied:
Right. Adding reset and wear level tests now.

geky reviewed (changes requested):
The implementation is looking good. I'm already impressed, I thought the timeline was too short for creating a new filesystem.

However, I have a problem with how this is being integrated.

1. **We don't need to introduce a new API**

   We have an established storage API (POSIX) that has already been adopted by users. We should not change directions without a good reason.

   If we **must** introduce a new storage API, the interface should be available for all storage options that support it. There’s no point in having multiple storage options if the users can’t transition between the storage options.

1. **StorageLite should not be tightly coupled to security**

   There is no technical reason StorageLite and security must be coupled. If it is an API problem, then we should fix our API so we can effectively add storage features without limiting users.

How can we move forward?

1. Integrate StorageLite into the POSIX API
1. Introduce POSIX secure layer

If POSIX API is proven to be inefficient/unworkable:
1. Create KVStore API
1. Integrate StorageLite into KVStore API
1. Integrate LittleFS and FAT into KVStore API (POSIX adapter?)
1. Introduce KVStore secure layer

These two options satisfy the above issues. I'm open to alternatives, but until the above issues are solved, merging this PR will be more disruptive than beneficial.

deepikabhavnani commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+
+int StorageLiteFS::mount(BlockDevice *bd)
+{
+    _bd = bd;
```
Error handling/checking of input parameters must be done, what if bd is NULL?

geky replied:
We don't check NULL pointers elsewhere. Worst case, code will hard fault, and the NULL dereference can easily be found with a debugger.

deepikabhavnani replied:
Yup, and it’s not good. Recently the code base I am touching I try to add where ever I can, hence pointed out. Can be found with debugger but not good for end product

geky replied:
Interesting, I think we disagree here. Should we also be asserting on misaligned pointers?

deepikabhavnani replied:
Error handling does not mean only asserting, but to actually avoid sudden crash and respond with proper errror message/code to help application recover (if possible). Misaligned address access depends on type of underlying block device. You have a good point, I will check how SD driver behaves as I don’t remember on top of my mind. But device should terminate gracefully even in case of unaligned access with error and not crash suddenly.

geky replied:
Every check we add increases code size. At some point we must trust the user is using our APIs correctly. I would consider a hard fault at the site of dereference a rather graceful error compared to other bugs they could run into with this API, ie buffer overruns in file_read.

kjbracey replied:
Strongly agree with @geky here. If you're going to check NULL pointers, you logically need to be checking for misaligned pointers, stale pointers, out-of-range pointers, corrupted data structures, so on and so forth.

If you are going to worry about the validity of pointers you need to be writing in something other than C or C++03.

Are you going to add `if (this == NULL)` checks to the top of every method too? You would logically have to in the model you advocate.

And what happens in practice if you do return an error code on an invalid pointer? The caller probably ignores it. So we get subtle effects like we've had with some `Mutex::lock()`s not working without anyone realising, causing hard to track-down problems.

Having said all that...

> the NULL dereference can easily be found with a debugger.

pyOCD still can't show a backtrace on a Cortex-M HardFault, as far as I know. It's been like that for a few years now. Rather undermines our argument, if pyOCD is our preferred debugger.

geky replied:
> pyOCD still can't show a backtrace on a Cortex-M HardFault

Oh, this would be a good thing to fix!

deepikabhavnani commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+
+int StorageLiteFS::remove(const char *filename)
+{
+    int ret = _stlite->remove((const uint8_t *) filename, strlen(filename) + 1);
```
Error handling/checking- filename

kjbracey replied:
Disagree, but the casts suggest a type problem - you should be passing filenames around as `char *` pointers, just like any string. If you need to cast to `uint8_t *` it should be lower level than this.

It this is converting to an opaque key, the opaque key's type should probably be `void *` to make caller's lives easier. And I wouldn't have thought you needed to include the NUL byte as part of the key, if you're giving explicit key lengths.

davidsaada replied:
Right. The idea was to have opaque keys. will change to `void *` and remove the null terminator.

deepikabhavnani commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+{
+    size_t file_size;
+    int ret = _stlite->get_file_size((const uint8_t *) name, strlen(name) + 1, file_size);
+    st->st_size = file_size;
```
Error handling/checking- name / st. Accessing NULL st here can cause crash.

deepikabhavnani commented on features/filesystem/storagelite/source/StorageLiteFS.cpp:
```diff
@@ -0,0 +1,320 @@
+    strcpy(handle->file_name, path);
+    handle->open_flags = flags;
+
+    *file = handle;
```
Error handling - user can pass NULL to file pointer.

kjbracey commented:
Agree that all the basic functions here - get/set/remove/get_file_size/get_next are perfectly standard filing system APIs and should be accessible via `FileSystem` and `FileHandle` (even if there is a more-specific `KVStore` API/wrapper).

There's no justification for not being able to do

    FILE *settings_file = fopen("/slite/settings", "r");
    if (settings_file) {
        fread(my_settings, 1, sizeof my_settings, settings_file);
        fclose(settings_file);
    }

Apps shouldn't need to be written to a special custom API to do something that simple.

Certainly calling this a "filing system" is rather confusing given that it is not actually an mbed OS filing system.

davidsaada commented:
@geky @kjbracey-arm Many of your observations were discussed in the design phase, so not sure they should also be discussed here as well. Still here's my take:
- StorageLite is not a file system: 
It is indeed not a file system. It is a key value store with security characteristics and additional features (such as factory reset). As such, it includes APIs to support key value store - e.g. get, set, remove etc.

- So why the POSIX APIs?
POSIX APIs were added later on upon requirement, for users who wanted a well known interface without porting issues.

- Now you got me confused. If it has POSIX APIs, it is a file system, isn't it?
Not quite. Many of the file system operations are not supported. Operations like seek, directory manipulations, renaming files and more. In fact, it only supports the open-read-close and open-write-close sequences. Anything else will fail.

- Oh. Why these restrictions?
StorageLite is designed to run on Flash devices. Although, like all other file systems, it runs on top of a block device, this block device must have flash attributes. Implementing a fully operational filesystem would have a great performance penalty if you only need key value store operations. In fact, we have a fully functional file system on to of Flash, called LittleFS. However, simulating a standard client provisioning action (writing 30 files of 512B each) takes less than a second in StorageLite and more than 8 seconds in LittleFS. In addition, Flash wear leveling in StorageLite is optimal - it only erases sectors after the whole file system space is exhausted, something that can't be achieved with a regular file system.

- OK, we can live with these restrictions. Why not use only the POSIX APIs and discard the other ones (get/set etc.)?
Not so fast. POSIX APIs don't offer the full functionality StorageLite offers, as they have a standard restricted interface. These include security features such as encryption and rollback protection on a per file basis, as well as factory reset. Security features are a must. Factory reset is required by the client. In addition, imitating the POSIX API behavior can also have a performance penalty. For instance, the open-read-close operation takes twice as much time comparing to the get operation. This is because when we open a file for reading, we need to check that the file exists on flash, aside from the read operation. Applications that don't need the standard APIs (like the client) would hence prefer using the get/set APIs over the POSIX ones.

- You can do better than that! Why not separate the additional features from the basic storage operations and implement it as separate layers?
We can, but again - it will have a great (negative) impact on performance and functionality. Factory reset, for one, will need to be implemented as a separate set of files (with different names), holding a lot of space in RAM and flash. As for the security features, trying to implement them (encryption & rollback protection) in the block device layer produced horrendous performance. Implementing these on top of the storage layer will mean that we will also have to separate security actions from storage actions, while in StorageLite they are combined. The file header, for instance, holds both security and storage fields. Separating those will waste storage space. Furthermore, this separation will lose the atomicity of operations. One example would be writing a large encrypted file. This is done in one shot in StorageLite. How would this be achieved after the separation? encrypt the entire file in memory? separate the file to smaller segments in storage? and if so, how do you link these segments or handle extreme cases (like reaching the end of our area if it happens)?

kjbracey commented:
There may a reason for an "optimised" custom API. If this were an entirely internal component written just for the purposes of mbed-client, and you were prepared to put the necessary work into its "storage abstraction layer", then it would be fine to stop there.

But that's still no reason to not support the standard API. There is nothing particularly unusual about a filing system with your restrictions, and there are many out there. And the majority of applications will work perfectly fine on them.

LTFS is a good example - it's a tape filing system, so has all the usual restrictions of tape - no random access, no reclaiming of space when a file is deleted until the tape is reformatted.

But it works fine for a lot of purposes with no custom code by just presenting itself as a standard POSIX filing system. Yes, dragging a load of files off a tape from Windows filer window can be inefficient, because it copies in alphabetical order rather than tape order, causing unnecessary seeking, but it works. And it doesn't preclude someone using a custom "tape-aware" program to do it better.

So even if the native interface remains as it stands, it should be an urgent priority to "complete the job" by providing an actual `StorageLiteFS` that implements a real `FilingSystem` on top of `StorageLite` so normal programs can just use it without having to invent their own "storage abstraction layer". This could be a separate wrapper - I doubt it would need to touch the internals.

And on atomicity, I was assuming the atomicity would be achieved simplistically by buffering the operation in the `FileHandle` - so the set actually happens on `close` and the get actually happens on `open`.

I'm not aware of a demand to manipulate excessively large items. If they were that large it would be equally unreasonable to expect them to be in a contiguous application buffer for the "set" or "get" as you currently do. Providing the POSIX interface could in some cases just move the required buffer from the app into the FileHandle, without increasing the total RAM.

geky commented:
@davidsaada, I like your summary. However, I disagree with some of your assumptions:

- > Implementing a fully operational filesystem [on flash] would have a great performance penalty if you only need key value store operations.

  That's just not true. StorageLite's optimizations come from minimizing erases and tracking file locations in RAM. Nothing related to the API.

  And comparing to LittleFS doesn't prove what's possible. LittleFS was designed for RAM/ROM consumption, not for speed.

- > Flash wear leveling in StorageLite is optimal - it only erases sectors after the whole file system space is exhausted, something that can't be achieved with a regular file system.

  That's not true either. Look at [SPIFFS](https://github.com/pellepl/spiffs). It is a fully features POSIX file system with static wear leveling.

  Keep in mind static wear leveling has its own set of issues in the form of wear amplification.

- > POSIX APIs don't offer the full functionality StorageLite offers, as they have a standard restricted interface.

  You can add features without giving up on the API. Look at all the block devices, they get along pretty well together.

- > imitating the POSIX API behavior can also have a performance penalty. For instance, the open-read-close operation takes twice as much time comparing to the get operation.

  So fix it. At this point you only need to authenticate the name, not also the rest of the file contents in open.

- > Factory reset, for one, will need to be implemented as a separate set of files (with different names), holding a lot of space in RAM and flash.

  Use a special name for files that share data. Or add support for reflink, as a COW filesystem LittleFS could handle this _very_ efficiently.

- > Trying to implement them (encryption & rollback protection) in the block device layer produced horrendous performance

  If you don't care about rollback protection, encrypting the block device was showing only a ~1.4x slowdown vs encrypting the filesystem. Not unreasonable given the ~0.8x RAM/ROM improvement.

- > The file header, for instance, holds both security and storage fields. Separating those will waste storage space.

  Separating things don't make them larger.

- > Furthermore, this separation will lose the atomicity of operations.

  Sounds like an API problem. Incremental writes to POSIX files would allow you to update the data atomically.

  If this needs to be applied to the KVStore API, consider adding user provided custom attributes. There's been discussions about adding these to LittleFS: https://github.com/geky/littlefs/issues/23#issuecomment-379128450. This would allow you to build features on top of storage.

- > One example would be writing a large encrypted file. This is done in one shot in StorageLite. How would this be achieved after the separation? encrypt the entire file in memory?

  Isn't this currently how StorageLite works? StorageLite files must always be stored entirely in memory. We should not be encouraging the use of StorageLite with large files.

  That's the biggest downside of the KVStore API, someone has to cough up the RAM somewhere.

davidsaada commented:
Pushed a commit with a few changes suggested in the comments. Mainly trying to handle build issues.
(Sorry for squashing, was unintentional).

dannybenor commented:
I strongly recommend to limit the PR review to code and not feature or product definitions.
Storage lite provides a limited POSIX API as required in the design after reviewed by tech leaders.
The limitations are not an accident but a result of design and analysis.

adbridge commented:
@geky @kjbracey-arm @deepikabhavnani  are you happy with the updates to this PR ?

davidsaada commented:
@adbridge From my POV, main gaps I still need to fill are the additional tests (resilience and wear level) and fixing the README according to the comments.

geky reviewed (changes requested):
- +1 for the additional tests (resilience and wear level)
- At minimum we still need a KVStore interface (abstract base class) for the new API that StorageLite implements.

And I still have problems with the design as stated above, this is being reviewed internally.

cmonr commented:
Marking as DNM due to ongoing internal review.

@simonfordarm

Commit:
Fix two issues in ExhaustibleBlockDevice:

1. Return errors when attempting to erase a worn out erase unit or write to it.
2. Support erase of more than one erase unit (in one call).

davidsaada commented:
(as it's a post release weekend)
/morph build

cmonr commented:
From our (@ARMmbed/mbed-os-maintainers) understanding, StorageLite requires a redesign before it can be reintroduced into Mbed OS. Since we don't support/allow PRs to remain open for development purposes, I'm going to close this PR.
